scripts/simulator/simulator_dao.js

```diff
@@ -0,0 +1,229 @@
+  console.log(`campaign ${campaignID} has totalVote=${totalVotes} voteCounts=${voteCounts} totalSuppy=${totalSupply}`)
+  let winOption = new BN(0)
+  let maxVotedCount = new BN(0)
+  for (let i = 0; i < voteCounts.length; i++) {
+    if (voteCounts[i].gt(maxVotedCount)) {
+      winOption = new BN(i + 1)
+      maxVotedCount = voteCounts[i]
+    } else if (voteCounts[i].eq(maxVotedCount)) {
+      winOption = new BN(0)
+    }
+  }
+
+  if (winOption.eq(new BN(0))) {
+    return [new BN(0), new BN(0), campaign.campaignType]
+  }
+
+  let votedPercentage = totalVotes.mul(precisionUnits).div(totalSupply)
+  if (campaign.minPercentageInPrecision.gt(votedPercentage)) {
+    return [new BN(0), new BN(0), campaign.campaignType]
+  }
+
+  let x = campaign.tInPrecision.mul(votedPercentage).div(precisionUnits)
+  if (!x.gt(campaign.cInPrecision)) {
+    let y = campaign.cInPrecision.sub(x)
+    if (maxVotedCount.mul(precisionUnits).lt(y.mul(totalVotes))) {
+      return [new BN(0), new BN(0), campaign.campaignType]
+    }
+  }
+  return [winOption, campaign.options[winOption.sub(new BN(1))], campaign.campaignType]
+}
+
+function addValueToDictionay (dic, key, value) {
+  if (key in dic) {
+    dic[key] = dic[key].add(value)
+  } else {
+    dic[key] = value
+  }
+}
+
+function subValueToDictionay (dic, key, value) {
+  if (key in dic) {
+    dic[key] = dic[key].sub(value)
+  } else {
+    dic[key] = value
+  }
+}
+
+module.exports.handlewithdraw = function (staker, reduceAmount, epoch, currentBlockTime) {
+  if (!(staker in numberVotes)) return
+  if (!(epoch in numberVotes[staker])) return
+  //if numberVotes contains value for state and epoch, that mean numVotes!=0
+  numVote = numberVotes[staker][epoch]
+  subValueToDictionay(totalEpochPoints, epoch, reduceAmount.mul(numVote))
+
+  if (!(epoch in epochCampaigns)) return
+  let campaignIds = epochCampaigns[epoch]
+
+  for (const campaignId of campaignIds) {
+    if (!(staker in stakerVotedOption)) continue
+    if (!(campaignId in stakerVotedOption[staker])) continue
+
+    votedOption = stakerVotedOption[staker][campaignId]
+
+    assert(campaignId in campaignData, 'campaign not exits')
+    let campaign = campaignData[campaignId]
+    // check if campaign has ended
+    if (campaign.endTimestamp < currentBlockTime) continue
+    campaign.campaignVoteData.totalVotes = campaign.campaignVoteData.totalVotes.sub(reduceAmount)
+    campaign.campaignVoteData.votePerOption[votedOption.sub(new BN(1))] = campaign.campaignVoteData.votePerOption[
+      votedOption.sub(new BN(1))
+    ].sub(reduceAmount)
+  }
+}
+
+module.exports.getStakerVoteCount = function (staker, epoch, totalStake) {
+  if (!(staker in numberVotes)) return zeroBN
+  if (!(epoch in numberVotes[staker])) return zeroBN
+  return numberVotes[staker][epoch]
+}
+
+module.exports.getTotalEpochPoints = function (epoch) {
+  if (!(epoch in totalEpochPoints)) return zeroBN
+  return totalEpochPoints[epoch]
+}
```

Comment:
I do it here
https://github.com/KyberNetwork/smart-contracts/blob/cc1ce65dfdeaaf3412a0a193387a634eae3553b6/test/sol6/daoFuzzTests.js#L407

Reply:
It might miss-match because totalEpochPoints is not total point of all campaign in 1 epoch. It is the total point of staker for reward distribution.

https://github.com/KyberNetwork/smart-contracts/blob/dc86610/contracts/sol6/Dao/KyberDao.sol#L170-L177

Instead I check the total point of each staker and total point in the simulator and in smart-contract
https://github.com/KyberNetwork/smart-contracts/blob/cc1ce65dfdeaaf3412a0a193387a634eae3553b6/test/sol6/daoFuzzTests.js#L407

Reply:
OK

This is great